File: midolmanj/src/test/java/com/midokura/midolman/monitoring/MonitoringTest.java
/*
* Copyright 2012 Midokura Europe SARL
*/
package com.midokura.midolman.monitoring;

import akka.actor.Actor;
import akka.testkit.TestActorRef;
import akka.testkit.TestKit;
import com.google.inject.AbstractModule;
import com.google.inject.Guice;
import com.google.inject.Injector;
import com.google.inject.Module;
import com.midokura.midolman.AbstractMidolmanTestCase;
import com.midokura.midolman.MidolmanTestCase;
import com.midokura.midolman.config.MidolmanConfig;
import com.midokura.midolman.guice.MidolmanModule;
import com.midokura.midolman.guice.actors.TestableMidolmanActorsModule;
import com.midokura.midolman.guice.cluster.MockClusterClientModule;
import com.midokura.midolman.guice.config.MockConfigProviderModule;
import com.midokura.midolman.guice.datapath.MockDatapathModule;
import com.midokura.midolman.guice.reactor.ReactorModule;
import com.midokura.midolman.guice.zookeeper.MockZookeeperConnectionModule;
import com.midokura.midolman.host.services.HostService;
import com.midokura.midolman.services.HostIdProviderService;
import com.midokura.midolman.services.MidolmanService;
import com.midokura.midolman.state.PortDirectory;
import com.midokura.midolman.vrn.AbstractVrnControllerTest;
import com.midokura.midonet.cluster.services.MidostoreSetupService;
import com.yammer.metrics.Metrics;
import com.yammer.metrics.core.Metric;
import com.yammer.metrics.core.MetricName;
import com.yammer.metrics.core.MetricsRegistry;
import org.apache.commons.configuration.HierarchicalConfiguration;
import org.junit.Before;
import org.openflow.protocol.OFPhysicalPort;
import org.openflow.protocol.OFPortStatus;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.testng.annotations.AfterMethod;
import org.testng.annotations.BeforeMethod;
import org.testng.annotations.Test;

import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.UUID;

import static java.lang.String.format;
import static org.hamcrest.CoreMatchers.is;
import static org.hamcrest.MatcherAssert.assertThat;
import static org.hamcrest.Matchers.*;

public class MonitoringTest extends AbstractMidolmanTestCase {

    private static final Logger log = LoggerFactory
            .getLogger(MonitoringTest.class);

    MonitoringAgent monitoringAgent;
    MetricsRegistry registry;


    @Test
    public void testVMMonitoring() {
        HierarchicalConfiguration config = fillConfig(new HierarchicalConfiguration());
        Injector injector = Guice.createInjector(getModulesAsJavaIterable(config));

        registry = Metrics.defaultRegistry();

        injector.getInstance(MidostoreSetupService.class).startAndWait();
        injector.getInstance(MidolmanService.class).startAndWait();
        injector.getInstance(MonitoringAgent.class).startMonitoring();
        Map<MetricName, Metric> allMetrics = registry.allMetrics();
        log.debug("Le ble");
        for (MetricName metricName: allMetrics.keySet()) {
            log.debug("METRIC --> {}", metricName);
        }
    }


   /* private List<Module> getModules() {
        List<Module> modulesList = new ArrayList<Module>() {{
                this.add(new MockConfigProviderModule(new HierarchicalConfiguration()));
                this.add(new MockDatapathModule());
                this.add(new MockZookeeperConnectionModule());
                this.add(new AbstractModule() {
                    @Override
                    protected void configure() {
                        bind(HostIdProviderService.class).toInstance(new HostIdProviderService() {
                            @Override
                            public UUID getHostId() {
                                return UUID.randomUUID();
                            }
                        });
                    }
                });
                 this.add(new ReactorModule());
                 this.add(new MockClusterClientModule());
                 scala.collection.mutable.Map<String,TestKit> probesByName = new scala.collection.mutable.HashMap<String,TestKit>();
                 scala.collection.mutable.Map<String,TestActorRef<Actor>> actorsByName = new scala.collection.mutable.HashMap<String, TestActorRef<Actor>>();
                 this.add(new TestableMidolmanActorsModule(probesByName, actorsByName));
                 this.add(new MidolmanModule());
            }
        };


        return modulesList;
    }

        */
}

File: midolmanj/src/test/scala/com/midokura/midolman/monitoring/MonitoringTest.scala
/*
* Copyright 2012 Midokura Europe SARL
*/
package com.midokura.midolman.monitoring

import com.midokura.midolman._
import com.yammer.metrics.Metrics
import com.yammer.metrics.core.Metric
import com.yammer.metrics.core.MetricName
import com.yammer.metrics.core.MetricsRegistry
import org.slf4j.Logger
import org.slf4j.LoggerFactory
import org.testng.annotations.Test
import java.util._
import org.hamcrest.MatcherAssert.assertThat
import collection.{mutable, immutable}
import scala.collection.JavaConversions._
import store.{MockStore, Store}
import com.midokura.midonet.cluster.data.{Bridge => ClusterBridge,
Ports => ClusterPorts}
import com.midokura.midonet.cluster.data.host.Host
import com.midokura.util.functors.Callback0
import com.midokura.tools.timed.Timed
import com.midokura.util.Waiters.waitFor
import org.apache.commons.configuration.HierarchicalConfiguration

object MonitoringTest {
}

class MonitoringTest extends MidolmanTestCase {

  private final val log: Logger = LoggerFactory.getLogger(classOf[MonitoringTest])

  var monitoringAgent: MonitoringAgent = null
  var registry: MetricsRegistry = null

  override protected def fillConfig(config: HierarchicalConfiguration): HierarchicalConfiguration = {
    config.setProperty("midolman.midolman_root_key", "/test/v3/midolman")
    config.setProperty("midolman.enable_monitoring", "true")

    config
  }

  @Test def testActualMonitoring {

    val jvmNames = immutable.Set(
      "OpenFileDescriptorCount", "ThreadCount", "FreePhysicalMemorySize",
      "TotalPhysicalMemorySize", "ProcessCPUTime", "TotalSwapSpaceSize", "MaxHeapMemory", "UsedHeapMemory",
      "CommittedHeapMemory", "AvailableProcessors", "FreeSwapSpaceSize", "SystemLoadAverage", "txBytes", "rxBytes",
      "rxPackets","txPackets")

    val store : Store = injector.getInstance(classOf[Store])
    val cassandraStore = store.asInstanceOf[MockStore]
    assertThat("Initial ports are empty", cassandraStore.getTargets.isEmpty);

    var called : Boolean = false;
    val host = new Host(hostId()).setName("myself")
    clusterDataClient().hostsCreate(hostId(), host)

    initializeDatapath()

    // make a bridge, and a port.
    val bridge = new ClusterBridge().setName("test")
    bridge.setId(clusterDataClient().bridgesCreate(bridge))
    val port = ClusterPorts.materializedBridgePort(bridge)


    port.setId(clusterDataClient().portsCreate(port))

    cassandraStore.subscribeToChangesRegarding(port.getId.toString, new Callback0 {
      def call() {
        called = true;
      }
    })

    clusterDataClient().hostsAddVrnPortMapping(host.getId, port.getId, "tapDevice")

    // make sure the monitoring agent receives data for the expected port.
    probeByName(MonitoringActor.Name).expectMsgType[DatapathController.PortStats].portID.toString should be (port.getId.toString)


    waitFor("Wait for the MidoReporter to write to Cassandra.",10000, 500,  new Timed.Execution[Boolean]{
      protected def _runOnce() {
          setCompleted(called);
          setResult(called)
      }
    })

    assertThat("Cassandra contains data about the port", called)

    // get all the metrics names
    registry = Metrics.defaultRegistry;
    val allMetrics: Map[MetricName, Metric] = registry.allMetrics
    val names = mutable.Set.empty[String]

    for (metricName <- allMetrics.keySet()) {
      names += metricName.getName
    }

    assertThat("The registry contains all the expected metrics.", (names.diff(jvmNames)).isEmpty )

    // clean stuff.
    registry.allMetrics().keys.foreach((arg: MetricName) => registry.removeMetric(arg))
  }



}